src/main/scala/com/scality/clueso/query/cache/SessionCacheManager.scala

```diff
@@ -0,0 +1,97 @@
+package com.scality.clueso.query.cache
+
+import java.util.Date
+import java.util.concurrent.atomic.AtomicReference
+
+import com.scality.clueso.CluesoConfig
+import com.scality.clueso.query.MetadataQueryExecutor.setupDf
+import com.typesafe.scalalogging.LazyLogging
+import org.apache.spark.sql.{DataFrame, SparkSession}
+import org.joda.time.DateTime
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.{Await, Future}
+
+object SessionCacheManager extends LazyLogging {
+  var bucketDfs = Map[String, AtomicReference[DataFrame]]()
+  var bucketUpdateTs = Map[String, DateTime]()
+
+  val setupDfLock = new scala.collection.parallel.mutable.ParHashSet[String]()
+
+  def getCachedBucketDataframe(spark: SparkSession, bucketName: String)(implicit config: CluesoConfig): DataFrame = {
+    // check if cache doesn't exist
+    val tableView = s"${new Date().getTime}_$bucketName"
+
+    if (!bucketDfs.contains(bucketName)) {
+      val bucketDf = setupDf(spark, config, bucketName)
+
+      if (acquireLock(bucketName)) {
+        bucketDf.createOrReplaceTempView(tableView)
+        bucketDf.sparkSession.catalog.cacheTable(tableView)
```

Comment:
does this actually trigger a read and a real cache?

Reply:
It does